modules/engine-prototype/src/main/scala/engine/Step.scala

```diff
@@ -0,0 +1,251 @@
+                        )
+                      )
+                    }
+                  ) *> false.pure[F]
+
+              }
+
+            def configureInst: F[Boolean] =
+              Execution.configureInst("F2").flatMap {
+                case Right(_) =>
+                  m.modify(
+                    Sequence.State.current.modify {
+                      case Pending.Standard(s) => Ongoing.Standard(
+                        F2.Ongoing.Standard(s.core, tcsConfigured = false, instConfigured = true)
+                      )
+                      case Ongoing.Standard(s) => Ongoing.Standard(
+                        F2.Ongoing.Standard(s.core, tcsConfigured = true, instConfigured = true)
+                      )
+                      case Failed.Standard(s) => Failed.Standard(
+                        F2.Failed.Standard(
+                          s.core,
+                          tcsFailed = true,
+                          instFailed = false,
+                          observationFailed = false
+                        )
+                      ) // Instrument already failed
+                    }
+                  ) *> true.pure[F]
+
+                case Left(_) =>
+                  m.modify(
+                    Sequence.State.current.modify {
+                      case Failed.Standard(s) => Failed.Standard(
+                        F2.Failed.Standard(
+                          s.core,
+                          tcsFailed = true,
+                          instFailed = true,
+                          observationFailed = false
+                        )
+                      )
+                      // TCS didn't fail yet
+                      case s => Failed.Standard(
+                        F2.Failed.Standard(
+                          s.core,
+                          tcsFailed = false,
+                          instFailed = true,
+                          observationFailed = false
+                        )
+                      )
+                    }
+                  ) *> false.pure[F]
+              }
+
+              // XXX: Check running status
+              both(configureTCS, configureInst).flatMap {
+              case (true, true) =>
+                  // XXX: Check running status
+                  Execution.observe.flatMap {
+                    case Some(fileId) => execute(m) // XXX: Shift current focus
+                    case None =>
+                      m.modify(
+                        Sequence.State.current.modify(s =>
+                          Failed.Standard(
+                            F2.Failed.Standard(
+                              s.core,
+                              tcsFailed = false,
+                              instFailed = false,
+                              observationFailed = true
+                            )
+                          )
+                        )
+                      ).void
+                  }
```

Comment:
In the case of failure the state is updated with the failed state and no recursive call is made so it stops. Again changing the Status is missing.